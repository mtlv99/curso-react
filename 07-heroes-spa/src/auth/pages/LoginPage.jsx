import React from 'react';

export const LoginPage = () => {
  console.log('');

  return (
    <h1>LoginPage</h1>
  );
};
