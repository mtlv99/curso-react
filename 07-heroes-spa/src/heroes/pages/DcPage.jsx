import React from 'react';

export const DcPage = () => {
  console.log('');

  return (
    <h1>DcPage</h1>
  );
};
