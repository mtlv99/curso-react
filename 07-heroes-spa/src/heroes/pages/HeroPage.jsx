import React from 'react';

// TODO: Result from search?
export const HeroPage = () => {
  console.log('');

  return (
    <h1>HeroPage</h1>
  );
};
