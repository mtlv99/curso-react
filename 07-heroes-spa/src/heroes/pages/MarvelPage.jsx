import React from 'react';

export const MarvelPage = () => {
  console.log('');

  return (
    <h1>MarvelPage</h1>
  );
};
