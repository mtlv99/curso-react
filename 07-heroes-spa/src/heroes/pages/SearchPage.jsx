import React from 'react';

export const SearchPage = () => {
  console.log('');

  return (
    <h1>SearchPage</h1>
  );
};
