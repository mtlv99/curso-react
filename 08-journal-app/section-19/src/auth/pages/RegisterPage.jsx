import React, { useMemo, useState } from 'react';
import { Link as RouterLink } from 'react-router-dom';
import {
  Alert,
  Button, Grid, Link, TextField, Typography,
} from '@mui/material';
import { Google } from '@mui/icons-material';
import { useDispatch, useSelector } from 'react-redux';
import { AuthLayout } from '../layout/AuthLayout';
import { useForm } from '../../hooks';
import { startCreatingUserWithEmailPassword } from '../../store/auth';

const formData = {
  displayName: 'Marco Leon',
  email: 'marco@example.com',
  password: '123456',
};

// La funcion que evalua acá, puede ser tan compleja como se desee (siempre que regrese un boolean).
const formValidations = {
  email: [(value) => value.includes('@'), 'El correo debe de tener una @'],
  password: [(value) => value.length >= 6, 'El password debe de tener más de 6 letras.'],
  displayName: [(value) => value.length >= 1, 'El nombre es obligatorio.'],
};

export const RegisterPage = () => {
  const dispatch = useDispatch();

  const { status, errorMessage } = useSelector((state) => state.auth);
  const isCheckingAuthentication = useMemo(() => status === 'checking', [status]);

  const [formSubmitted, setFormSubmitted] = useState(false);

  const {
    formState, displayName, email, password, onInputChange,
    isFormValid, displayNameValid, emailValid, passwordValid,
  } = useForm(formData, formValidations);


  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmitted(true);

    if (!isFormValid) return;

    dispatch(startCreatingUserWithEmailPassword(formState));
  };

  return (
    <AuthLayout title="Registro">
      <form onSubmit={onSubmit}>
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Nombre"
              type="text"
              placeholder="Nombre Completo"
              fullWidth
              name="displayName"
              value={displayName}
              onChange={onInputChange}
              // Acá se usa double negative debido a que en el mismo property viene
              // el mensaje de error
              error={!!displayNameValid && formSubmitted}
              helperText={displayNameValid}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Correo"
              type="email"
              placeholder="marco@example.com"
              fullWidth
              name="email"
              value={email}
              onChange={onInputChange}
              // Acá se usa double negative debido a que en el mismo property viene
              // el mensaje de error
              error={!!emailValid && formSubmitted}
              helperText={emailValid}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Contraseña"
              type="password"
              placeholder="Contraseña"
              fullWidth
              name="password"
              value={password}
              onChange={onInputChange}
              // Acá se usa double negative debido a que en el mismo property viene
              // el mensaje de error
              error={!!passwordValid && formSubmitted}
              helperText={passwordValid}
            />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>
            {/* eslint-disable-next-line no-extra-boolean-cast */}
            <Grid item xs={12} display={!!errorMessage ? '' : 'none'}>
              <Alert severity="error">
                {errorMessage}
              </Alert>
            </Grid>

            <Grid item xs={12}>
              <Button
                type="submit"
                variant="contained"
                fullWidth
                disabled={isCheckingAuthentication}
              >
                Crear cuenta
              </Button>
            </Grid>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/login">
              ¿Ya tienes cuenta?
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
