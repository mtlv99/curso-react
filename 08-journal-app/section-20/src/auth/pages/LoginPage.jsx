import React, { useMemo, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { Link as RouterLink } from 'react-router-dom';
import {
  Alert, Button, Grid, Link, TextField, Typography,
} from '@mui/material';
import { Google } from '@mui/icons-material';
import { useForm } from '../../hooks';
import { AuthLayout } from '../layout/AuthLayout';
import { startEmailSignIn, startGoogleSignIn } from '../../store/auth';


const formData = {
  email: 'marco@example.com',
  password: '123456',
};

const formValidations = {
  email: [(value) => value.includes('@'), 'El correo debe de tener una @'],
};


export const LoginPage = () => {
  const dispatch = useDispatch();

  const { status, errorMessage } = useSelector((state) => state.auth);

  // Cuidado con el valor por defecto de useForm! Luego de agregar el useEffect al useForm
  // que detecta cambios en el state inicial, si se pasa un objeto directamente
  // como atributo, cada vez que se rerenderiza el componente, estaría creando un
  // nuevo objeto con una nueva posición en memoria, por lo que estaría creando
  // ciclo/cycle infinito de rerenders!
  const {
    email, password, onInputChange, emailValid, passwordValid, isFormValid,
  } = useForm(formData, formValidations);

  const [formSubmitted, setFormSubmitted] = useState(false);


  const isAuthenticating = useMemo(() => status === 'checking', [status]);

  const onSubmit = (event) => {
    event.preventDefault();
    setFormSubmitted(true);

    if (!isFormValid) return;

    // No olvidar llamar la función en vez de solo pasarla!
    const result = dispatch(startEmailSignIn({ email, password }));
  };

  const onGoogleSignIn = (event) => {
    event.preventDefault();
    setFormSubmitted(true);
    // No olvidar llamar la función en vez de solo pasarla!
    dispatch(startGoogleSignIn());
  };

  return (
    <AuthLayout title="Login">
      <form onSubmit={onSubmit} className="animate__animated animate__fadeIn animate__faster">
        <Grid container>
          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Correo"
              type="email"
              placeholder="marco@example.com"
              fullWidth
              name="email"
              value={email}
              onChange={onInputChange}
              // Leer comentario de RegisterPage.jsx
              error={!!emailValid && formSubmitted}
              helperText={emailValid}
            />
          </Grid>

          <Grid item xs={12} sx={{ mt: 2 }}>
            <TextField
              label="Contraseña"
              type="password"
              placeholder="Contraseña"
              fullWidth
              name="password"
              value={password}
              onChange={onInputChange}
              // Leer comentario de RegisterPage.jsx
              error={!!passwordValid && formSubmitted}
              helperText={passwordValid}
            />
          </Grid>

          <Grid container spacing={2} sx={{ mb: 2, mt: 1 }}>

            {/* eslint-disable-next-line no-extra-boolean-cast */}
            <Grid item xs={12} display={!!errorMessage ? '' : 'none'}>
              <Alert severity="error">
                {errorMessage}
              </Alert>
            </Grid>

            <Grid item xs={12} sm={6}>
              <Button type="submit" variant="contained" fullWidth disabled={isAuthenticating}>
                Login
              </Button>
            </Grid>
            <Grid item xs={12} sm={6}>
              <Button variant="contained" fullWidth onClick={onGoogleSignIn} disabled={isAuthenticating}>
                <Google />
                <Typography sx={{ ml: 1 }}>Google</Typography>
              </Button>
            </Grid>
          </Grid>

          <Grid container direction="row" justifyContent="end">
            <Link component={RouterLink} color="inherit" to="/auth/register">
              Crear una cuenta
            </Link>
          </Grid>
        </Grid>
      </form>
    </AuthLayout>
  );
};
