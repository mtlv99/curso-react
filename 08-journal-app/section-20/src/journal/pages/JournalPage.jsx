import React from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { IconButton, Typography } from '@mui/material';
import { AddOutlined } from '@mui/icons-material';
import { JournalLayout } from '../layout/JournalLayout';
import { NoteView, NothingSelectedView } from '../views';
import { startNewNote } from '../../store/journal';

export const JournalPage = () => {
  const dispatch = useDispatch();
  const { isSaving, activeNote } = useSelector((state) => state.journal);

  const onClickNewNote = () => {
    dispatch(startNewNote());
  };

  // Con el `component` attribute se transforma en un h1 pero conserva el tamaño de un parrafo normal.
  // Con el `variant` attribute se transforma en un h1, y aplica el tamaño de un h1 definido por defecto en Material UI.
  // <Typography component="h1">Journal Page</Typography>
  // <Typography variant="h1">Journal Page</Typography>
  return (
    <JournalLayout>

      {/* eslint-disable-next-line no-extra-boolean-cast */}
      {!!activeNote ? <NoteView /> : <NothingSelectedView /> }

      {/* <Typography>Lorem ipsum dolor sit amet consectetur adipisicing elit. Ut, dolore doloremque. Rem aperiam, ipsum eius soluta, harum provident quia vero commodi hic veniam doloribus veritatis suscipit unde tempore, dolorum dolor.</Typography> */}

      {/* NothingSelected */}
      {/* <NothingSelectedView /> */}

      {/* NoteView */}
      {/* <NoteView /> */}

      <IconButton
        size="large"
        sx={{
          color: 'white',
          backgroundColor: 'error.main',
          ':hover': { backgroundColor: 'error.main', opacity: 0.9 },
          position: 'fixed',
          right: 50,
          bottom: 50,
        }}
        onClick={onClickNewNote}
        disabled={isSaving}
      >
        <AddOutlined sx={{ fontSize: 30 }} />
      </IconButton>

    </JournalLayout>
  );
};
