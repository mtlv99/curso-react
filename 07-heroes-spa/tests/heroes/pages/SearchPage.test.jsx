import { fireEvent, render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import { SearchPage } from '../../../src/heroes/pages';

const mockedUseNavigate = jest.fn();

// Ver Navbar.test.jsx para más detalles de este mock.
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: () => mockedUseNavigate,
}));

// TODO: Mock por defecto, para solucionar error de queryString.
jest.mock('query-string', () => ({ parse: jest.fn().mockReturnValue({ q: 'batman' }) }));

describe('Pruebas en <SearchPage />', () => {
  beforeEach(() => jest.clearAllMocks());

  // TODO: Este test está incorrecto porque no tiene los valores por defecto, pero el error me impide hacerlo.
  test('debe de mostrarse correctamente con valores por defecto', () => {
    // jest.mock('query-string', () => ({ parse: jest.fn().mockReturnValueOnce({ q: '' }) }));
    const { container } = render(
      // Si vemos un useLocation() o useNavigate() en el componente, significa que se ocupa el MemoryRouter
      <MemoryRouter>
        <SearchPage />
      </MemoryRouter>,
    );
    // screen.debug();

    expect(container).toMatchSnapshot();
  });

  test('debe de mostrar a Batman y el input con el valor del queryString', () => {
    // jest.mock('query-string', () => ({ parse: jest.fn().mockReturnValueOnce({ q: 'batman' }) }));

    render(
      <MemoryRouter initialEntries={['/search?q=batman']}>
        <SearchPage />
      </MemoryRouter>,
    );

    const input = screen.getByRole('textbox');
    expect(input.value).toBe('batman');

    // screen.debug();

    const img = screen.getByRole('img');
    expect(img.src).toContain('/assets/heroes/dc-batman.jpg');

    const notFound = screen.getByLabelText('not-found');
    const defaultMessage = screen.getByLabelText('default-message');

    expect(notFound.style.display).toBe('none');
    expect(defaultMessage.style.display).toBe('none');
  });


  // TODO: el test está incorrecto, no he logrado hacer que un mock devuelva diferentes valores.
  // Esto por el problema del mock de queryString.
  test('debe de mostrar un error si no se encuentra el hero (batman123)', () => {
    render(
      <MemoryRouter initialEntries={['/search?q=batman123']}>
        <SearchPage />
      </MemoryRouter>,
    );
    const notFound = screen.getByLabelText('not-found');
    // expect(notFound.style.display).toBe('');
    expect(notFound.style.display).toBe('none');
  });


  test('debe de llamar el navigate() para ir a la pantalla nueva', () => {
    const inputValue = 'superman';
    render(
      <MemoryRouter initialEntries={['/search']}>
        <SearchPage />
      </MemoryRouter>,
    );
    const input = screen.getByRole('textbox');
    fireEvent.change(input, { target: { name: 'searchText', value: inputValue } });
    const form = screen.getByRole('form');
    fireEvent.submit(form);
    expect(mockedUseNavigate).toHaveBeenCalledWith(`?q=${inputValue}`);
  });
});
