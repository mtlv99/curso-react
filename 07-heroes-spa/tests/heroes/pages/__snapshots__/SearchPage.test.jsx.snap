// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <SearchPage /> debe de mostrarse correctamente con valores por defecto 1`] = `
<div>
  <h1>
    Search
  </h1>
  <hr />
  <div
    class="row"
  >
    <div
      class="col-5"
    >
      <h4>
        Looking for a hero?
      </h4>
      <hr />
      <form
        aria-label="form"
      >
        <input
          autocomplete="off"
          class="form-control"
          name="searchText"
          placeholder="Search a hero..."
          type="text"
          value="batman"
        />
      </form>
    </div>
    <div
      class="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        aria-label="default-message"
        class="alert alert-primary animate__animated animate__fadeIn"
        style="display: none;"
      >
        Search a hero...
      </div>
      <div
        aria-label="not-found"
        class="alert alert-danger animate__animated animate__fadeIn"
        style="display: none;"
      >
        No results found for 
        <b>
          batman
        </b>
        .
      </div>
      <div
        class="col"
      >
        <div
          class="card animate__animated animate__fadeIn"
        >
          <div
            class="row no-gutters"
          >
            <div
              class="col-4"
            >
              <img
                alt="Batman"
                class="card-img"
                src="/assets/heroes/dc-batman.jpg"
              />
            </div>
            <div
              class="col-8"
            >
              <div
                class="card-body"
              >
                <h5
                  class="card-title"
                >
                  Batman
                </h5>
                <p
                  class="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  class="card-text"
                >
                  <small
                    class="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                >
                  Leer m√°s...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
